File: apps/js/src/lib/token.ts
import jwt from "jsonwebtoken";

export const TTL = "24h";

function secret(): string {
	return process.env["JWT_SECRET"] ?? "";
}

export function sign(idUser: number): string {
	return jwt.sign({}, secret(), {
		algorithm: "HS256",
		subject: String(idUser),
		expiresIn: TTL,
	});
}

export function parse(raw: string): number {
	// reject algorithm confusion
	const claims = jwt.verify(raw, secret(), { algorithms: ["HS256"] });

	if (typeof claims === "string" || claims.sub === undefined) {
		throw new Error("token has no subject");
	}

	return Number(claims.sub);
}

File: apps/go/internal/token/token.go
package token

import (
	"os"
	"strconv"
	"time"

	"github.com/golang-jwt/jwt/v5"
)

const TTL = 24 * time.Hour

func secret() []byte {
	return []byte(os.Getenv("JWT_SECRET"))
}

func Sign(idUser int64) (string, error) {
	now := time.Now()

	return jwt.NewWithClaims(jwt.SigningMethodHS256, jwt.RegisteredClaims{
		Subject:   strconv.FormatInt(idUser, 10),
		IssuedAt:  jwt.NewNumericDate(now),
		ExpiresAt: jwt.NewNumericDate(now.Add(TTL)),
	}).SignedString(secret())
}

func Parse(raw string) (int64, error) {
	claims := &jwt.RegisteredClaims{}

	_, err := jwt.ParseWithClaims(raw, claims, func(*jwt.Token) (any, error) {
		return secret(), nil
	}, jwt.WithValidMethods([]string{jwt.SigningMethodHS256.Alg()})) // reject algorithm confusion
	if err != nil {
		return 0, err
	}

	return strconv.ParseInt(claims.Subject, 10, 64)
}
